components/file_system/src/io_stats/proc.rs

```diff
@@ -0,0 +1,251 @@
+        let pid = thread::process_id();
+        let tid = thread::thread_id();
+        ThreadID {
+            pid,
+            tid,
+            proc_reader: None,
+        }
+    }
+
+    fn fetch_io_bytes(&mut self) -> Option<IOBytes> {
+        if self.proc_reader.is_none() {
+            let path = PathBuf::from("/proc")
+                .join(format!("{}", self.pid))
+                .join("task")
+                .join(format!("{}", self.tid))
+                .join("io");
+            match File::open(path) {
+                Ok(file) => {
+                    self.proc_reader = Some(BufReader::new(file));
+                }
+                Err(e) => {
+                    warn!("failed to open proc file: {}", e);
+                }
+            }
+        }
+        if let Some(ref mut reader) = self.proc_reader {
+            reader
+                .seek(std::io::SeekFrom::Start(0))
+                .map_err(|e| {
+                    warn!("failed to seek proc file: {}", e);
+                })
+                .ok()?;
+            let mut io_bytes = IOBytes::default();
+            for line in reader.lines() {
+                let line = line
+                    .map_err(|e| {
+                        // ESRCH 3 No such process
+                        if e.raw_os_error() != Some(3) {
+                            warn!("failed to read proc file: {}", e);
+                        }
+                    })
+                    .ok()?;
+                if line.len() > 11 {
+                    let mut s = line.split_whitespace();
+                    if let (Some(field), Some(value)) = (s.next(), s.next()) {
+                        if field.starts_with("read_bytes") {
+                            io_bytes.read = u64::from_str(value).ok()?;
+                        } else if field.starts_with("write_bytes") {
+                            io_bytes.write = u64::from_str(value).ok()?;
+                        }
+                    }
+                }
+            }
+            Some(io_bytes)
+        } else {
+            None
+        }
+    }
+}
+
+struct LocalIOStats {
+    id: ThreadID,
+    io_type: IOType,
+    last_flushed: IOBytes,
+}
+
+impl LocalIOStats {
+    fn current() -> Self {
+        LocalIOStats {
+            id: ThreadID::current(),
+            io_type: IOType::Other,
+            last_flushed: IOBytes::default(),
+        }
+    }
+}
+
+#[derive(Default)]
+struct AtomicIOBytes {
+    read: AtomicU64,
+    write: AtomicU64,
+}
+
+impl AtomicIOBytes {
+    fn load(&self, order: Ordering) -> IOBytes {
+        IOBytes {
+            read: self.read.load(order),
+            write: self.write.load(order),
+        }
+    }
+
+    fn fetch_add(&self, other: IOBytes, order: Ordering) {
+        self.read.fetch_add(other.read, order);
+        self.write.fetch_add(other.write, order);
+    }
+}
+
+/// Flushes the local I/O stats to global I/O stats.
+#[inline]
+fn flush_thread_io(sentinel: &mut LocalIOStats) {
```

Comment:
How often is this happend. Seem that this function involves IO read

Reply:
It involves syscall, but no IO. As procfs is a virtual file that is redirected to kernel memory. Plus the metrics flusher only calls it once every 10s. For the overhead, you can check out the benchmark results in https://github.com/tikv/tikv/pull/11873#issuecomment-1017080413.

Reply:
got